File: packages/fixtures/src/reference-graph.ts
import { parseRef } from "./refs.js";
import type { Diagnostic, ReferenceGraph, ReferenceGraphNode, ExtractedReference } from "./types.js";

export interface GraphBuildEntry {
  ref: string;
  value: unknown;
  resolved: boolean;
  refs: ReadonlyArray<ExtractedReference>;
}

export function buildReferenceGraph(
  entries: ReadonlyArray<GraphBuildEntry>,
  diagnostics: readonly Diagnostic[] = [],
): ReferenceGraph {
  const nodes = new Map<string, ReferenceGraphNode>();
  const edges: ReferenceGraph["edges"] = [];

  for (const entry of entries) {
    addNode(nodes, entry.ref, entry.resolved);
    if (!entry.resolved) continue;

    for (const ref of entry.refs) {
      addNode(nodes, ref.ref, false);
      edges.push({ from: entry.ref, to: ref.ref, fieldPath: ref.fieldPath });
    }
  }

  const resolved = new Set(entries.filter((entry) => entry.resolved).map((entry) => entry.ref));
  for (const node of nodes.values()) {
    if (resolved.has(node.ref)) node.resolved = true;
  }

  return { nodes, edges, diagnostics: [...diagnostics] };
}

function addNode(nodes: Map<string, ReferenceGraphNode>, ref: string, resolved: boolean): void {
  const existing = nodes.get(ref);
  if (existing) {
    if (resolved) existing.resolved = true;
    return;
  }

  let type = "<malformed>";
  let id = ref;
  try {
    const parsed = parseRef(ref);
    type = parsed.type;
    id = parsed.id;
  } catch {
    // Malformed refs stay visible as unresolved nodes.
  }

  nodes.set(ref, { ref, type, id, resolved });
}

File: python/fixtures/src/mirk/fixtures/reference_graph.py
"""Build the reference graph from loaded entries.

A malformed ref stays visible as a node with type `<malformed>` rather than
being dropped, so a broken pack shows the broken edge instead of a gap.
"""

from __future__ import annotations

from collections.abc import Sequence
from typing import Any, NamedTuple

from .errors import FixtureError
from .refs import parse_ref
from .types import Diagnostic, ExtractedReference, ReferenceGraph, ReferenceGraphNode

__all__ = ["GraphBuildEntry", "build_reference_graph"]


class GraphBuildEntry(NamedTuple):
    ref: str
    value: Any
    resolved: bool
    refs: Sequence[ExtractedReference]


def build_reference_graph(
    entries: Sequence[GraphBuildEntry],
    diagnostics: Sequence[Diagnostic] = (),
) -> ReferenceGraph:
    nodes: dict[str, ReferenceGraphNode] = {}
    edges: list[Any] = []

    for entry in entries:
        _add_node(nodes, entry.ref, entry.resolved)
        if not entry.resolved:
            continue
        for reference in entry.refs:
            _add_node(nodes, reference["ref"], False)
            edges.append(
                {"from": entry.ref, "to": reference["ref"], "fieldPath": reference["fieldPath"]}
            )

    resolved = {entry.ref for entry in entries if entry.resolved}
    for node in nodes.values():
        if node["ref"] in resolved:
            node["resolved"] = True

    return {"nodes": nodes, "edges": edges, "diagnostics": list(diagnostics)}


def _add_node(nodes: dict[str, ReferenceGraphNode], ref: str, resolved: bool) -> None:
    existing = nodes.get(ref)
    if existing is not None:
        if resolved:
            existing["resolved"] = True
        return

    type_name = "<malformed>"
    identifier = ref
    try:
        parsed = parse_ref(ref)
    except FixtureError:
        pass
    else:
        type_name = parsed.type
        identifier = parsed.id

    nodes[ref] = {"ref": ref, "type": type_name, "id": identifier, "resolved": resolved}
